File: abc317/c/main.cpp
#include <bits/stdc++.h>
using namespace std;
#define rep(i, n) for (int i = 0; i < n; i++)
#define mxin(a, b) (a = max(a, b))
using ll = long long;
using vl = vector<ll>;
using vvl= vector<vl>;
using pll = pair<ll, ll>;

ll max_path = 0;

void dfs(int cur, ll cur_cost, vector<vector<pll>> &adj, vector<bool> &visited) {
    visited[cur] = true;
    mxin(max_path, cur_cost);
    for (pll next : adj[cur]) {
        if (!visited[next.first]) {
            dfs(next.first, cur_cost + next.second, adj, visited);
        }
    }
    visited[cur] = false;
}


int main() {
    int N, M;
    cin >> N >> M;
    vector<vector<pll>> adj(N);
    rep(i, M) {
        ll A, B, C;
        cin >> A >> B >> C;
        adj[A-1].emplace_back(B-1, C);
        adj[B-1].emplace_back(A-1, C);
    }

    vector<bool> visited(N, false);
    rep(i, N) {
        dfs(i, 0, adj, visited);
    }

    cout << max_path << endl;

    return 0;
}

File: abc317/c/main.py
# DFS関数の定義
def dfs(cur, cur_cost, adj, visited):
    global max_path  # グローバル変数を使用
    max_path = max(max_path, cur_cost)  # 経路の最大値を更新

    visited[cur] = True  # 現在のノードを訪問済みにする

    for neighbor, cost in adj[cur]:
        if not visited[neighbor]:  # 未訪問の隣接ノードに対して
            dfs(neighbor, cur_cost + cost, adj, visited)  # DFSを再帰的に実行

    visited[cur] = False  # 現在のノードの訪問状態を戻す

# main関数
if __name__ == '__main__':
    N, M = map(int, input().split())  # ノード数Nと辺の数M
    adj = [[] for _ in range(N)]  # 隣接リスト
    max_path = 0  # 経路の最大値

    # 辺の情報を読み取り、隣接リストを作成
    for _ in range(M):
        A, B, C = map(int, input().split())
        adj[A-1].append((B-1, C))
        adj[B-1].append((A-1, C))  # 無向グラフなので、双方向に追加

    visited = [False] * N  # 訪問状態を保存するリスト

    # 各ノードからDFSをスタート
    for start in range(N):
        dfs(start, 0, adj, visited)

    print(max_path)  # 経路の最大値を出力
